### ISSUE TITLE
Have questions on implementation of LabelSmoothingLoss and ignore_index

### ISSUE THREAD

**Issue by sonsus** [OpenNMT/OpenNMT-py#1463]
Below is the source code of <code>LabelSmoothingLoss</code> of this package(openNMT). I guess that ignore_index is default -100 and not sure what does <code>one_hot[self.ignore_index] = 0 </code> is doing here.

1. Am I getting it right about default value of <code>ignore_index = -100 </code>as it is in <code>nn.NLLLoss() </code> and it is implemented here for the same purpose? 

2. If so, how does this work? I tried to read both source code of<code>nn.Module</code> and this but I failed to follow the lines. 

```python
class LabelSmoothingLoss(nn.Module):
    """
    With label smoothing,
    KL-divergence between q_{smoothed ground truth prob.}(w)
    and p_{prob. computed by model}(w) is minimized.
    """
    def __init__(self, label_smoothing, tgt_vocab_size, ignore_index=-100):
        assert 0.0 < label_smoothing <= 1.0
        self.ignore_index = ignore_index
        super(LabelSmoothingLoss, self).__init__()

        smoothing_value = label_smoothing / (tgt_vocab_size - 2)
        one_hot = torch.full((tgt_vocab_size,), smoothing_value)
        one_hot[self.ignore_index] = 0 # here is the thing. 
        self.register_buffer('one_hot', one_hot.unsqueeze(0))

        self.confidence = 1.0 - label_smoothing
```

**Comment by sonsus** [comment]
OK nobody wants to answer this from discuss-pytorch nor here. Thanks.

**Comment by guillaumekln** [comment]
A few lines above, `ignore_index` is set to the padding index: https://github.com/OpenNMT/OpenNMT-py/blob/e8622eb5c6117269bb3accd8eb6f66282b5e67d9/onmt/utils/loss.py#L35

Is that was you were looking for?

**Comment by sonsus** [comment]
Not exactly. Setting padding index to be ignored when computing the loss is plausible enough. What I want to know is that what's being done inside the <code>LabelSmoothingLoss.\_\_init\_\_()</code>. Ignore idx is default -100 and what would things go like when <code>one_hot[-100] = 0</code> as shown in the permalink to the source below? Wouldn't the index -100 will be treated same as <code>one_hot[vocab_length-100]=0</code> which might corresponds to another word in the vocab? I'm not sure it has special meaning or convention that I don't know or it just simply assumes that vocab_length>100 and -100's position corresponds to <code>'\<pad\>'</code> in one_hot vector.

https://github.com/OpenNMT/OpenNMT-py/blob/e8622eb5c6117269bb3accd8eb6f66282b5e67d9/onmt/utils/loss.py#L199

**Comment by sonsus** [comment]
And sorry if I hurt your feelings. I was bad about not using permalink properly to relieve your load for understanding and answering this question. Thanks for giving me a good example of using permalink. I didn't noticed how to use this functionality before.

**Comment by guillaumekln** [comment]
So after some research, this default value does come from `NLLLoss` but in the PyTorch implementation it is not used in a tensor indexing expression, meaning that the default negative value is to not ignore any index. The implementation has checks like `cur_target == ignore_index` where `cur_target` is always positive.

In the context of the OpenNMT-py implementation, the default value is indeed incorrect based on what you said:

> Ignore idx is default -100 and what would things go like when one_hot[-100] = 0 as shown in the permalink to the source below? Wouldn't the index -100 will be treated same as one_hot[vocab_length-100]=0 which might corresponds to another word in the vocab?

I suggest sending a PR to set the default value to `None` and add a condition around `one_hot[self.ignore_index] = 0`.

**Comment by sonsus** [comment]
I should go thru the lines for doing that with details. Thx for kind reply! I really appreciate =]
